lib: lte_link_control: Set default system mode to LTE-M + NB-IoT + GNSS

tokangas opened the pull request:
Changed default system mode to LTE-M + NB-IoT + GNSS. Earlier, the default was LTE-M + GNSS, but because `CONFIG_LTE_NETWORK_USE_FALLBACK` was also enabled, the functionality was similar. Instead of the LTE link control library, the modem now decides which system to use depending on the LTE mode preference.

Removed system mode configuration from samples and applications which do not need it. The system mode configuration is kept for special cases like the NIDD sample, which requires NB-IoT.

NordicBuilder commented:
## Test specification

### [CI/Jenkins/NRF](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-13280/1/)
- Integration Platforms

### [CI/Jenkins/integration](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-13280/1/)
| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-ci-nrfconnect-boot-fw-update | X | | |
| test-fw-nrfconnect-nrf-iot_cloud | X | | |
| test-fw-nrfconnect-nrf-iot_lwm2m | X | | |
| test-fw-nrfconnect-nrf-iot_mosh | X | | |
| test-fw-nrfconnect-nrf-iot_positioning | X | | |
| test-fw-nrfconnect-nrf-iot_samples | X | | |
| test-fw-nrfconnect-nrf-iot_thingy91 | X | | |
| test-fw-nrfconnect-nrf-iot_zephyr_lwm2m | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-13280/1/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

tokangas commented:
I'm not sure if `CONFIG_LTE_MODE_PREFERENCE_AUTO` is the best default for mode preference or not. Another possibility would be `CONFIG_LTE_MODE_PREFERENCE_LTE_M_PLMN_PRIO`, which prefers LTE-M, but not over PLMN selection (modem prioritizes the home network).

The problem with `CONFIG_LTE_MODE_PREFERENCE_AUTO` is that once modem selects NB-IoT, it tends to keep using it as long as NB-IoT is available (also after modem is put into offline mode and back to online). We may want to prioritize LTE-M instead.

tokangas commented on samples/cellular/lwm2m_client/prj.conf:
```diff
@@ -59,7 +59,6 @@ CONFIG_STREAM_FLASH=y
 
 # LTE link control
 CONFIG_LTE_LINK_CONTROL=y
```
@SeppoTakalo @jheiskan81 The LwM2M client sample has a separate overlay for NB-IoT. If LTE-M has not been available, NB-IoT has eventually been tried anyway, even without the overlay (because of `CONFIG_LTE_NETWORK_USE_FALLBACK`). The question is, do we want to keep LTE-M and NB-IoT separate, or should we remove the NB-IoT overlay and enable both LTE-M and NB-IoT in the default configuration as I have done now? If we want to support both with the same configuration, `CONFIG_COAP_INIT_ACK_TIMEOUT_MS` needs to be increased.

tokangas replied:
I restored the LwM2M client configuration back to as it was. The default configuration is LTE-M only and there's an overlay to switch to NB-IoT. As discussed with @jheiskan81, implementation of dynamic ack timeout is ongoing, after which there's no longer need to use different configurations for LTE-M and NB-IoT.

gregersrygg reviewed (commented):
The asset_tracker_v2 change is fine by us. The rest is owned by others, so I'll leave the approval for other reviewers.

trantanen reviewed (changes requested):
No real comments. Release note entry is the only thing missing at this point.

tokangas commented:
> I'm not sure if `CONFIG_LTE_MODE_PREFERENCE_AUTO` is the best default for mode preference or not. Another possibility would be `CONFIG_LTE_MODE_PREFERENCE_LTE_M_PLMN_PRIO`, which prefers LTE-M, but not over PLMN selection (modem prioritizes the home network).

Changed the default LTE mode preference to `CONFIG_LTE_MODE_PREFERENCE_LTE_M_PLMN_PRIO`. Also updated the LTE mode preference dependencies, it makes no sense to select any other preference than `CONFIG_LTE_MODE_PREFERENCE_AUTO` when only LTE-M or NB-IoT is enabled, although this is allowed by the AT command specification.

tokangas commented:
Deprecated the `CONFIG_LTE_NETWORK_USE_FALLBACK` Kconfig option.

tokangas commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -443,13 +443,17 @@ Modem libraries
       Use the :c:func:`lte_lc_connect` function instead.
     * The :c:func:`lte_lc_init_and_connect_async` function is deprecated.
       Use the :c:func:`lte_lc_connect_async` function instead.
+    * The :kconfig:option:`CONFIG_LTE_NETWORK_USE_FALLBACK` Kconfig option is deprecated.
```
Do we want to already remove the link? The option is deprecated for now, but not removed yet.

peknis replied:
Oh, my bad then. I thought you took it out already.

lemrey commented on lib/lte_link_control/Kconfig:
```diff
@@ -278,9 +274,7 @@ config LTE_RAI_REQ_VALUE
 config LTE_NETWORK_USE_FALLBACK
-	bool "Use fallback network mode if selected mode fails"
-	default y if !(LTE_NETWORK_MODE_LTE_M_NBIOT || LTE_NETWORK_MODE_LTE_M_NBIOT_GPS || \
-		       LTE_NETWORK_MODE_DEFAULT)
+	bool "Use fallback network mode if selected mode fails [DEPRECATED]"
```
Let this `select DEPRECATED` so that `cmake` emits a warning when it is set.

trantanen commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -443,13 +443,17 @@ Modem libraries
       Use the :c:func:`lte_lc_connect` function instead.
     * The :c:func:`lte_lc_init_and_connect_async` function is deprecated.
       Use the :c:func:`lte_lc_connect_async` function instead.
+    * The :kconfig:option:`CONFIG_LTE_NETWORK_USE_FALLBACK` Kconfig option is deprecated.
```
I would move this to 3rd bullet below network mode and LTE mode preference as this is tight to those changes.

trantanen commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -443,13 +443,17 @@ Modem libraries
 
   * Updated:
 
+    * The default network mode to :kconfig:option:`CONFIG_LTE_NETWORK_MODE_LTE_M_NBIOT_GPS`.
```
```suggestion
    * The default network mode to :kconfig:option:`CONFIG_LTE_NETWORK_MODE_LTE_M_NBIOT_GPS` from :kconfig:option:`CONFIG_LTE_NETWORK_MODE_LTE_M_GPS`.
```

trantanen commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -443,13 +443,17 @@ Modem libraries
   * Updated:
 
+    * The default network mode to :kconfig:option:`CONFIG_LTE_NETWORK_MODE_LTE_M_NBIOT_GPS`.
+    * The default LTE mode preference to :kconfig:option:`CONFIG_LTE_MODE_PREFERENCE_LTE_M_PLMN_PRIO`.
```
```suggestion
    * The default LTE mode preference to :kconfig:option:`CONFIG_LTE_MODE_PREFERENCE_LTE_M_PLMN_PRIO` from :kconfig:option:`CONFIG_LTE_MODE_PREFERENCE_AUTO`.
```

trantanen commented on lib/lte_link_control/Kconfig:
```diff
@@ -292,6 +289,7 @@ config LTE_NETWORK_USE_FALLBACK
 	  instruct the modem on which network to use. This is more efficient
 	  both in terms of time to connection establishment and energy
 	  consumption.
+	  This Kconfig option is deprecated. Use CONFIG_LTE_MODE_PREFERENCE_* instead.
```
```suggestion
	  This Kconfig option is deprecated. Use the :kconfig:option:`CONFIG_LTE_NETWORK_MODE_LTE_M_NBIOT` or :kconfig:option:`CONFIG_LTE_NETWORK_MODE_LTE_M_NBIOT_GPS` instead. In addition, you can control the priority between LTE-M and NBIoT with :kconfig:option:`CONFIG_LTE_MODE_PREFERENCE` Kconfig option.
```
